### PR TITLE
File System Object Structure and Database

### PR DESCRIPTION
**Pull request by QuentCov**
Joint branch work from Collin and myself. Contains the FileSystemObject classes, the Database object and the database scripts.

### CODE REVIEW

**Comment by LostLaplace** [comment]
Why are File and FileObject separate classes? They seem like duplicates to me.

**Comment by QuentCov** [comment]
FileSystemObject is the abstract class that File and Directory are derived from, so that we can make the tree hold FileSystemObjects and thus contain all Files and Directories.

**Comment by LostLaplace** [comment]
Yes, but if you look in code/database.py there is a FileObject. There are also a few lines of code that aren't part of a class. How are we going to run them? I think there should be a Database class that initializes the engine and and binding that needs to be done in SQLAlchemy.

File: code/database.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy.ext.declarative import declarative_base
+import datetime
+import hashlib
+from FileSystemObject import File, Directory
+
+
+
+
+engine = create_engine('sqlite:///SyncOrSwimDB.db')
```
Comment (LostLaplace) [inline code/database.py:11]: These lines of code are not part of any class. Should these be in the Database class?

File: code/database.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	def __init__(self, path, lastModified, deleted, toEncrypt, lastSync, children):
+		self.path = path
+		self.lastModified = lastModified
+		self.deleted = deleted
+		self.toEncrypt = toEncrypt
+		self.lastSync = lastSync
+		self.children = children
+
+
+def retrieve(path):
```
Comment (LostLaplace) [inline code/database.py:83]: These functions are also not part of any class.

**Comment by QuentCov** [comment]
Oh, I misinterpreted your first comment. I think FileObject was something that Collin was using to test the database, but it appears that they didn't come out any different from each other. I can't really say what the methods in database.py are for, but the three variables at the top appear to be globals for use in all the classes in that file.

File: code/database.py (hunk 2 above)
Comment (cvictor2) [reply code/database.py:83]: I tested them from the command line. I don't think they need to be in any class. To run one of the functions you need to 
import database
database.function()
and it should run whatever function you called from the database moduer

**Comment by LostLaplace** [comment]
We'll have to figure that out. I'm not sure how globals work when you import the classes into a different file... If it works it works though I won't complain.

I guess what I'm saying is it may be better practice to, instead of using globals, have those variables as member variables of a class that are set in the constructor.

**Comment by cvictor2** [comment]
FileObject is a mapping class. To store something in the database I convert a File or Directory into a FileObject or DirectoryObject, which then stores that in the database. For retrieval, it gets the FileObject or DirectoryObject from the database and converts that into a File or Directory object. I don't think I can store a File or Directory directly into the database, so I used an intermediate class to convert Files and Directories from/into database entries.

**Comment by LostLaplace** [comment]
Ok. I think that's fair. Do we think it's a good idea to change the constructor for the FileObject and DirectoryObject to just accept a File or Directory object instead of each attribute individually?

File: code/database.py (hunk 2 above)
Comment (LostLaplace) [reply code/database.py:83]: Ok, so what's happening here is that to python the database.py file is considered a separate module. So when you run import on it it imports the symbol table of database.py (including its functions and variables) into the top modules symbol table under the object "database". I'm not sure if there are any downsides to this or not, so we may just leave it.

**Comment by cvictor2** [comment]
I could probably change that, but will that really speed anything up or change anything? The constructor will still need to get each individual attribute.

**Comment by LostLaplace** [comment]
It'll make it easier for the code that is pushing or pulling out of the database. They won't have to decompose all of the attributes to create a FileObject. They can just do:
`myFile = File(initiallize file...)
myFileObject = FileObject(myFile)`

And be done with it.

**Comment by cvictor2** [comment]
Yes, but it would still be decomposing all of the attributes in the constructor. Would this do anything besides make the code look nicer?

**Comment by QuentCov** [comment]
Making the code look good is not a bad thing to do if its this simple.

**Comment by LostLaplace** [comment]
I mean, if we decompose the object in the constructor then we only need to do it once. If we do it elsewhere we may be decomposing multiple times in different parts of the code.

**Comment by cvictor2** [comment]
ok, I'll make sure to change that. I also need to change how directories are retrieved, and probably add a delete function. Is there any other functionality I should add?

**Comment by LostLaplace** [comment]
Not that I can think of.

**Comment by QuentCov** [comment]
Just as a review of what the Database module does from my understanding: It pulls, stores, and now deletes items in the database. Can it update items?

**Comment by LostLaplace** [comment]
If my understanding of sqlalchemy is correct, "committing" the same object into the database that you pulled out earlier should update it. You've done more research Collin. What do you think?

**Comment by QuentCov** [comment]
According to stack overflow: 
user.no_of_logins += 1
session.commit()
will take the no_of_logins column of whatever row user represents and increase it by one. So it would appear that commiting does do updates for us.

**Comment by LostLaplace** [comment]
Nice. This is so much better than all that SQL we wrote in 156H am I right?

**Comment by cvictor2** [comment]
I use session.merge(obj), during the store function, so if that object is already in the database it will update it, otherwise it will add it to the database.

**Comment by cvictor2** [comment]
Just added the delete function and changed the constructor. I also added a convert function to both FileObject and DirectoryObject to easily convert them to File and Directory objects.

**Comment by QuentinCovert** [comment]
As discussed in the meeting minutes, I'm going to add a parameter for the folder holding the object. (probably not actually.)

**Comment by BurningSkyMK1** [comment]
If I'm correct, it was stated that the database had a list of root directories Watchman was watching. If there could be a function that could provide this list, ideally in the form of absolute paths, that'd be great. This is (obviously) directed toward the Database module.

**Comment by cvictor2** [comment]
The database keeps track of relative paths, I'm adding functionality to retrieve the roots, but its the relative paths. I can add an extra table if we want to keep track of the absolute paths of the roots if thats what we need.

**Comment by cvictor2** [comment]
Also, would this retrieve the roots as Directory objects, or just the absolute paths of the roots?

**Comment by cvictor2** [comment]
I looked through the specs sheet, but I did not find anything about the database keeping a list of root directories. I will add a new table for this, and create a method to retrieve a list of these

**Comment by BurningSkyMK1** [comment]
Any files or directories under "root" directories can have paths relative to the root directory. What I need is for the root directory to contain an absolute path (of the system it's on). Preferably, somewhere they're would be a list of roots so I can populate them in our program and in the GUI.

**Comment by cvictor2** [comment]
Should that be included in the root directory's Directory object, or should I write a function to return a list of all the absolute paths for each root?

**Comment by LostLaplace** [comment]
We could just define the root directories with an absolute path and the normal ones with a relative path. We would need to add a isRoot flag to the Directory class.

**Comment by LostLaplace** [comment]
The trouble I see with that is what if you want to mirror it to a different path on a different system? On one machine I might want to mirror /home/cameron/myDocuments, but on a different machine I might want to sync /home/cjohnson/myDocuments

**Comment by LostLaplace** [comment]
I think this is where the local config file comes into play. Each machine will need to have it's own way to map root directories to absolute paths on that specific system.

**Comment by cvictor2** [comment]
Yeah, that would work, although I suggest for each root we make the parent the absolute path to the parent directory, and leave the relative path the same

**Comment by LostLaplace** [comment]
That's a good idea too.

**Pull request closed without merging**